olp-cpp-sdk-core/include/olp/core/thread/SyncQueue.h

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (C) 2019 HERE Europe B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+#pragma once
+
+#include <condition_variable>
+#include <deque>
+#include <mutex>
+#include <queue>
+
+#include <olp/core/CoreApi.h>
+
+namespace olp {
+namespace thread {
+
+/**
+ * @brief The SyncQueue class is a thread safe container adapter.
+ *
+ * @tparam T queue item to be stored inside the buffer.
+ * @tparam Container The type of the underlying container that should mimic the
+ * API of std::queue providing the following functions:
+ * - empty()
+ * - push(T&&)
+ * - push(const T&)
+ * - front()
+ * - pop()
+ */
+template <typename T, typename Container>
+class CORE_API SyncQueue final {
```

Comment:
What about Front() method, which just retrieves top element without pulling it?  I guess this method will be helpful for the users of this class.

Reply:
Means two or more concurrent threads can access the same element at the same time? This is super tricky and dangerous. I would not provide this possibility.

Reply:
Ok, let's stick to the YAGNI principle.